torch/distributed/checkpoint/_v2/_checkpointing.py

```diff
@@ -0,0 +1,409 @@
+
+
+@dataclass
+class RankInfo:
+    global_rank: int
+    global_world_size: int
+    role_rank: int
+    role_world_size: int
+    role_index: int
+    role_replica_count: int  # to support PAFT with HSDP
+
+
+@dataclass
+class CheckpointContext:
+    step: int
+    extra_context: dict[str, Any]
+
+
+class Checkpointer(abc.ABC):
+    """
+    This is a checkpointing solution to store and load models with parallelism at scale.
+
+    Note: If you working with single rank models and do not need asynchronous checkpointing, we recommend
+    using `torch.save` and `torch.load` for its simplicity.
+
+    This class provides extension points for users to customize individual components as they
+    need.
+
+    .. warning::
+        This feature is experimental and subject to removal/change.
+
+    """
+
+    @abc.abstractmethod
+    def save(
+        self,
+        state_dict: dict[str, Any],
+        context: CheckpointContext,
+        root_dir: str,
+        use_cached_manifest: bool = False,
+    ) -> Optional[tuple[Future[None], Future[None]]]:
+        """
+        Save a checkpoint to a given path on storage. This optionally saves the metadata aggregated
+        across all ranks to make resharding on load efficient and generic.
+
+        Using module names as top level keys is optional but recommended. See the below example for
+        typical usage with a state dict containing module names as keys and module state dicts as
+        values.
+            {
+                "model": model_state_dict,
+                "optimizer": optimizer_state_dict,
+                "dataloader": dataloader_state_dict,
+                "metrics": metrics_state_dict,
+                "extra_state": extra_state_dict,
+            }
+
+        We expect most users to use a variant of async checkpointing implementation and so the API is
+        designed to be easy to use for this case. A synchronous checkpointing implementation
+        can still be implemented by on top of this API. EG: synchronous version can return a tuple of
+        two futures which are set with a result in the save() method.
+
+        Args:
+            state_dict (dict[str, Any]): The state_dict to save.
+            context (CheckpointContext): The context to save the checkpoint.
+            root_dir (str): The path to save the checkpoint.
+            use_cached_manifest (bool): Whether to use cached manifest. (Default: ``False``). If
+                use_cached_manifest is True, it will use the compute the manifest to save the checkpoint.
+                Otherwise, it will compute the manifest and save the checkpoint. If the checkpoint is saved
+                successfully, it will return a tuple of two futures. The first future is a future for the
+                manifest and the second future is a future for the checkpoint.
+
+        Returns:
+            tuple[Future, Future]: A tuple of two futures. The first future can be awaited on for completion
```

Comment:
conflicting definition?

above:
" The first future is a future for the
                manifest and the second future is a future for the checkpoint."